app/controllers/troubleshooting_controller.rb

```diff
@@ -0,0 +1,173 @@
+  before_action :load_item
+  before_action :check_visibility
+
+  # Display options for troubleshooting.
+  def show
+    @item_type = @item.class.base_class.model_name.human
+    @page_subtitle = ts("Troubleshoot %{type}",
+                        type: @item_type)
+  end
+
+  # Perform the desired troubleshooting actions.
+  def update
+    actions = params.fetch(:actions, []).map(&:to_s).reject(&:blank?)
+
+    disallowed = actions - @allowed_actions
+
+    if disallowed.any?
+      flash[:error] =
+        ts("The following actions aren't allowed: %{actions}.",
+           actions: disallowed.to_sentence)
+      redirect_to troubleshooting_path
+      return
+    end
+
+    flash[:notice] = []
+    flash[:error] = []
+
+    (@allowed_actions & actions).each do |action|
+      send(action)
+    end
+
+    # Make sure that we don't show blank errors.
+    flash[:notice] = nil if flash[:notice].blank?
+    flash[:error] = nil if flash[:error].blank?
+
+    redirect_to item_path
+  end
+
+  protected
+
+  # Calculate the permitted actions based on the item type and whether the
+  # current user is an admin. This is used in the show action to figure out
+  # which options to display, and in the update action to figure out which
+  # actions to perform.
+  #
+  # In order to work properly, this needs to return a list of strings, and each
+  # string needs to be the name of an instance method in this class. To make
+  # the list of options display properly (with a name and a description), each
+  # of these names also needs a corresponding title and description entry in
+  # the i18n scope en.troubleshooting.show.
+  def allowed_actions
+    if @item.is_a?(Tag) && logged_in_as_admin?
+      %w[fix_associations fix_counts fix_meta_tags update_tag_filters reindex_tag]
+    elsif @item.is_a?(Tag)
+      %w[fix_associations fix_counts fix_meta_tags]
+    elsif @item.is_a?(Work)
+      %w[update_work_filters reindex_work]
+    end
+  end
+
+  # Let the views use these two path methods.
+  helper_method :item_path, :troubleshooting_path
+
+  # The path to view the item. Ideally we could just do redirect_to @item, but
+  # unfortunately we're dealing with tags, which have numerous subclasses.
+  def item_path
+    if @item.is_a?(Tag)
+      tag_path(@item)
+    else
+      polymorphic_path(@item)
+    end
+  end
+
+  # The path to view the troubleshooting page for this item. Again, this is
+  # necessary because tags have a lot of subclasses and need special handling.
+  def troubleshooting_path
+    if @item.is_a?(Tag)
+      tag_troubleshooting_path(@item)
+    else
+      polymorphic_path([@item, :troubleshooting])
+    end
+  end
+
+  # Load the @item based on params. Results in a 404 error if the item in
+  # question can't be found, and a 500 error if there's an unknown type. Also
+  # sets the variable @allowed_actions.
+  def load_item
+    if params[:tag_id]
+      @item = Tag.find_by_name(params[:tag_id])
```

Comment:
Using `find_by!` instead of `find_by_name` will break the page for all tags with special characters:
https://github.com/otwcode/otwarchive/blob/ff332552fadbec7161f5e0bd5a4c904f333ba88e/app/models/tag.rb#L600-L611

Reply:
I forgot we did this. Wish we used percent-encoding instead of a custom scheme.

Reply:
Maybe it deserves an issue? Unless there's a specific reason to want to use those custom escape characters. (I guess they're easier to understand than hex?)

Reply:
I don't know of a specific reason for the scheme, but other reviewers might.

I think we're unlikely to change it at this point because of backwards compatibility, though.

Reply:
I'm pretty sure it was just to make it easier for folks to read/memorize, yeah. I wouldn't be opposed to having a discussion about changing it.